### ISSUE TITLE
Nested Arguments

### ISSUE THREAD

**Issue by niqdev** [ghostdogpr/caliban#507]
Hi,

I'm trying to re-implement a very simplified version of the GitHub [api](https://developer.github.com/v4/explorer) to play with the pagination following the [spec](https://relay.dev/graphql/connections.htm), but I'm not entirely sure how to represent the nested arguments. Could you please give me some hints?

Thanks in advance for the help!

---

Query example
```graphql
query {
  user(login: "ghostdogpr") {
    id
    repositories(first: 2, after: "Y3Vyc29yOnYyOpHOA9J0lQ==") {
      totalCount
      edges {
        node {
          id
          name
        }
        cursor
      }
      pageInfo {
        endCursor
        startCursor
        hasPreviousPage
        hasNextPage
      }
    }
  }
}
```

This is a gist of my current schema
```scala
object queries {

  object arguments {
    final case class UserArg(login: NonEmptyString)
    final case class ForwardPaginationArg(
      first: Offset,
      after: Option[Cursor]
    )
  }

  final case class Queries[F[_]](
    // I don't know how to declare the pagination arguments for the nested "repositories" node
    user: UserArg => F[Option[UserNode]],
    // this works, but only because it's a root node and not what I want
    repositories: ForwardPaginationArg => F[RepositoryConnection]
  )
}
```
If needed, I can provide more details about the schema or the implementation, but I think they are neglectable

**Comment by ghostdogpr** [comment]
You can use arguments at any level of the schema: you just have the field return a function from the Arg case class to the result type. The same thing you did works whatever the level 😄

**Comment by niqdev** [comment]
Whoo, thanks for the super fast replay!

🤔 Not sure i've understood correctly, should i change the query or the schema?

```scala
final case class Queries[F[_]](
   // ??? should i change the query with something like this?
    user: UserArg => ForwardPaginationArg => F[Option[UserNode]] => F[RepositoryConnection]
  )

// ??? or should i change the schema like this
final case class UserNode(
    id: NodeId,
    name: NonEmptyString,
    createdAt: Instant,
    updatedAt: Instant,
    repositories: ForwardPaginationArg => F[RepositoryConnection]
  ) extends Node
```

Could you give me a super simple example if i'm on the wrong track? Thanks 😊 

---

This should give a full picture of what I got so far
```scala
object schema extends CommonSchema with CommonArgBuilder {

  @newtype case class Offset(value: NonNegLong)
  @newtype case class NodeId(value: Base64String)
  @newtype case class Cursor(value: Base64String)

  @GQLInterface
  sealed trait Node {
    def id: NodeId
  }

  final case class UserNode(
    id: NodeId,
    name: NonEmptyString,
    createdAt: Instant,
    updatedAt: Instant,
    //repository: Repository,
    repositories: RepositoryConnection
  ) extends Node

  // TODO add issue|issues
  final case class RepositoryNode(
    id: NodeId,
    name: NonEmptyString,
    url: String Refined Url,
    isFork: Boolean,
    createdAt: Instant,
    updatedAt: Instant
  ) extends Node

  final case class RepositoryConnection(
    edges: List[RepositoryEdge],
    nodes: List[RepositoryNode],
    pageInfo: PageInfo,
    totalCount: NonNegLong
  )

  final case class RepositoryEdge(
    cursor: Cursor,
    node: RepositoryNode
  )

  final case class PageInfo(
    hasNextPage: Boolean,
    hasPreviousPage: Boolean,
    startCursor: Cursor,
    endCursor: Cursor
  )
}
```

**Comment by ghostdogpr** [comment]
The second one ☝️ 

There is an example that shows nested effects here: https://github.com/ghostdogpr/caliban/tree/master/examples/src/main/scala/caliban/optimizations

Also this article might be of interest: https://medium.com/@ghostdogpr/graphql-in-scala-with-caliban-part-2-c7762110c0f9

**Comment by niqdev** [comment]
Awesome, thanks a lot for the help and for the great lib!

One last question, is there any plan in future to add *pagination* and *filter* support directly in caliban, as an extra module? Or it's too specific cos might vary case by case, hence it's out of scope?

**Comment by ghostdogpr** [comment]
No immediate plan because each case is specific.

There was some interest from a few contributors to add some helpers for the [Relay](https://relay.dev/) format so this might happen at some point.

**Comment by niqdev** [comment]
Cool, I'm working on a standalone [pr](https://github.com/niqdev/scala-fp/pull/85) in my spare time (i haven't pushed the latest changes and I'm heavily iterating over it) with the goal to abstract it to create a lib, e.g. `caliban-extras` and I'd like to do the same for the filter using [droste](https://github.com/higherkindness/droste) most likely.

I'm far from having anything close to be presented, but if you are interested, I could share it once it's ready and I'd be happy to contribute to this project - if it makes sense of course!

**Comment by ghostdogpr** [comment]
Sure, feel free to drop by the Discord later if you want to discuss this with me and other contributors/users.
